gitcmd/gitcmd.go

```diff
@@ -57,11 +71,58 @@ func (a GitHubPATAuther) InsertAuth(url string) string {
 	return url
 }
 
+// GitHubAppAuther authenticates using a GitHub App instead of a PAT.
+type GitHubAppAuther struct {
+	ClientID       string
+	InstallationID int64
+	PrivateKey     string // The GitHub App's private key (PEM format)
+}
+
+func (a GitHubAppAuther) InsertAuth(url string) string {
+	token, err := a.getInstallationToken()
+	if err != nil {
+		log.Printf("Failed to get GitHub App installation token: %v", err)
+		return url
+	}
+	if after, found := stringutil.CutPrefix(url, githubPrefix); found {
+		return fmt.Sprintf("https://x-access-token:%v@github.com/%v", token, after)
+	}
+	return url
+}
+
+func (a GitHubAppAuther) InsertHTTPAuth(req *http.Request) {
+	token, err := a.getInstallationToken()
+	if err != nil {
+		log.Printf("Failed to get GitHub App installation token: %v", err)
+		return
+	}
+	req.Header.Set("Authorization", "token "+token)
+}
+
+func (a GitHubAppAuther) getInstallationToken() (string, error) {
+	// Generate a JWT using the private key
+	jwt, err := githubutil.GenerateJWT(a.ClientID, a.PrivateKey)
+	if err != nil {
+		return "", err
+	}
+
+	// Exchange JWT for an installation token
+	token, err := githubutil.FetchInstallationToken(jwt, a.InstallationID)
+	if err != nil {
+		return "", err
+	}
+	return token, nil
```

Comment:
Can/should we be using the same token for more than one request? Is it just difficult to deal with expiration so we're making one per request for now? Worth a followup issue if so, IMO.

Reply:
Add a comment explaining this, and linking the issue?